materialize-kafka/src/transactor.rs

```diff
@@ -0,0 +1,211 @@
+use anyhow::{Context, Result};
+use avro::{encode, encode_key};
+use bytes::Bytes;
+use proto_flow::materialize::{
+    request::{Open, Store},
+    response::{Acknowledged, Flushed, Opened, StartedCommit},
+    Response,
+};
+use rdkafka::{
+    error::KafkaError,
+    producer::{BaseRecord, Producer},
+    types::RDKafkaErrorCode,
+    util::Timeout,
+};
+use tuple::Element;
+
+use crate::{
+    binding_info::{get_binding_info, BindingInfo},
+    configuration::{EndpointConfig, MessageFormat},
+    Input, Output,
+};
+
+pub async fn run_transactions(mut input: Input, mut output: Output, open: Open) -> Result<()> {
+    let spec = open
+        .materialization
+        .expect("must have a materialization spec");
+
+    let config: EndpointConfig = serde_json::from_str(&spec.config_json)?;
+    let producer = config.to_producer()?;
+
+    output.send(Response {
+        opened: Some(Opened {
+            runtime_checkpoint: None,
+        }),
+        ..Default::default()
+    })?;
+
+    let bindings = get_binding_info(
+        &spec.bindings,
+        &config.message_format,
+        config.schema_registry.as_ref(),
+    )
+    .await?;
+
+    let mut key_buf = Vec::new();
+    let mut payload_buf = Vec::new();
+
+    loop {
+        let request = match input.read()? {
+            Some(req) => req,
+            None => return Ok(()),
+        };
+
+        if request.flush.is_some() {
+            output.send(Response {
+                flushed: Some(Flushed { state: None }),
+                ..Default::default()
+            })?;
+            tracing::debug!("wrote Flushed");
+        } else if let Some(store) = request.store {
+            let mut msg = msg_for_store(
+                &config.message_format,
+                &mut key_buf,
+                &mut payload_buf,
+                store,
+                &bindings,
+            )?;
+
+            loop {
+                match producer.send(msg) {
+                    Ok(_) => {
+                        break;
+                    }
+                    Err((err, original_message)) => match err {
+                        KafkaError::MessageProduction(RDKafkaErrorCode::QueueFull) => {
+                            tracing::debug!("polling producer since queue is full");
+                            producer.poll(Timeout::Never);
+                            msg = original_message;
```

Comment:
is this line necessary? 🤔

Reply:
`producer.send` takes ownership of its `msg` argument, so without this line the borrow checker complains that on subsequent iterations of the retry loop `msg` has been moved. The `producer.send` error response gives you back the message it took ownership of and this reassignment of that to the original `msg` lets the next round of the loop use that one (which is the same one, really) rather than the moved variable.